overlord/managers_test.go

```diff
@@ -5849,3 +5862,164 @@ type: kernel`
 		"snap_try_kernel": "",
 	})
 }
+
+type gadgetUpdatesSuite struct {
+	baseMgrsSuite
+}
+
+var _ = Suite(&gadgetUpdatesSuite{})
+
+func (ms *gadgetUpdatesSuite) SetUpTest(c *C) {
+	ms.baseMgrsSuite.SetUpTest(c)
+
+	bloader := boottest.MockUC16Bootenv(bootloadertest.Mock("mock", c.MkDir()))
+	bootloader.Force(bloader)
+	ms.AddCleanup(func() { bootloader.Force(nil) })
+
+	restore := release.MockOnClassic(false)
+	ms.AddCleanup(restore)
+
+	mockServer := ms.mockStore(c)
+	ms.AddCleanup(mockServer.Close)
+
+	st := ms.o.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// setup model assertion
+	model := ms.brands.Model("can0nical", "my-model", modelDefaults, map[string]interface{}{
+		"gadget": "pi",
+	})
+	devicestatetest.SetDevice(st, &auth.DeviceState{
+		Brand:  "can0nical",
+		Model:  "my-model",
+		Serial: "serialserial",
+	})
+	err := assertstate.Add(st, model)
+	c.Assert(err, IsNil)
+}
+
+func (ms *gadgetUpdatesSuite) makeMockedDev(c *C, structureName string) {
+	// mock /dev/disk/by-label/{structureName}
+	byLabelDir := filepath.Join(dirs.GlobalRootDir, "/dev/disk/by-label/")
+	err := os.MkdirAll(byLabelDir, 0755)
+	c.Assert(err, IsNil)
+	// create fakedevice node
+	err = ioutil.WriteFile(filepath.Join(dirs.GlobalRootDir, "/dev/fakedevice0p1"), nil, 0644)
+	c.Assert(err, IsNil)
+	// and point the mocked by-label entry to the fakedevice node
+	err = os.Symlink(filepath.Join(dirs.GlobalRootDir, "/dev/fakedevice0p1"), filepath.Join(byLabelDir, structureName))
+	c.Assert(err, IsNil)
+
+	// mock /proc/self/mountinfo with the above generated paths
+	ms.AddCleanup(osutil.MockMountInfo(fmt.Sprintf("26 27 8:3 / %[1]s/%s rw,relatime shared:7 - vfat %[1]s/dev/fakedevice0p1 rw", dirs.GlobalRootDir, structureName)))
+
+	// and mock the mount point
+	err = os.MkdirAll(filepath.Join(dirs.GlobalRootDir, structureName), 0755)
+	c.Assert(err, IsNil)
+}
+
+// tsWithoutReRefresh removes the re-refresh task from the given taskset.
+//
+// It assumes that re-refresh is the last task and will fail if that is
+// not the case.
+//
+// This is needed because settle() will not converge with the re-refresh
+// task because re-refresh will always be in doing state.
```

Comment:
also the fact that we have issues/bugs reported about the re-refresh task getting stuck and taking a long time spinning it's wheels I think this is papering over the fact that we have a real bug, so at the very least this should somewhere have a TODO to fix it for reasl

Reply:
I agree that this is not ideal:/ The problem only manifests itself when there is a refresh that involves a reboot. Then then we have two tasks is "Doing" state (auto-connect that waits for the reboot and check-rerefresh) and settle does not converge. I could probably also try to just run Ensure() by hand until it reaches "Doing" in auto-connect instead of doing this and calling settle(). 

AFAIK we don't have a kernel refresh manage test,(we have remodel ones) those would also trigger this issue. I will add kernel refresh tests too as part of the "upgrade from snap without kernel.yaml to kernel.yaml" later too fwiw.

Reply:
hmm are you sure we don't already have similar tests? We have for example TestInstallKernelSnap20UpdatesBootloaderEnv, and the non-20 version in this same file. Granted, these tests uses `snapstate.InstallPath` instead of `snapstate.Update`, not sure if it matters much, but those tests also simulate reboots and things... perhaps also using `Update()` is why this test is not working right with the re-refresh test while the InstallPath doesn't trigger that.

Reply:
It's exactly the difference between Install/Update that is the issue here. Only Update() adds the check-rerefresh task.

Reply:
to be clear the issue here is not a real bug but that settle is based on the ensure next time stabilizing but the re-refresh check uses:

https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L2894

which I would expect moves it each time, so we never converge.

This is an expedient solution,  a more elegant one maybe would be to have a variant of Settle that ends if ensure next time is stable or in the future by a value larger than some threshold, and then we would mock the rerefresh interval to something large and distinct from practical wait time even on slow systems. I don't think we should block on this but something to consider, maybe leave a TODO here.

Reply:
Maybe I didn't ask this clearly enough, but do we _need_ to use `snapstate.Update` here? It seems to me that the additional things that `Update` does aren't paramount to this particular unit test? If indeed the re-refresh task is the primary difference from `Update` then for the purposes of testing gadget updates, shouldn't `snapstate.InstallPath` be sufficient which hits the same codepath?

I guess I just feel uneasy about having a test which creates a task set, then goes digging around in the result and removing things which we don't actually test or care about here - why not just make the test about what we care about?

Also I must admit I have always found re-refreshes very confusing and so my lack of familiarity with them is probably at least part to blame for feeling uneasy about this sort of change.

Reply:
This particular test would not snapstate.Update() but there will be a test with a update from a non-kernel-ref gadget to a gadget that uses kernel-refs to test the various scenarios where we go from an old style DTB to this new style. Anyway, I will look into the suggestion from Samuele to see if that fixes things.